content/stdout-vs-stderr.md

```diff
@@ -0,0 +1,2077 @@
+
+0\. Standard Input (`stdin`): for reading input.  
+1\. Standard Output (`stdout`): for writing conventional output.  
+2\. Standard Error (`stderr`): for printing diagnostic or error messages.
+
+Here is an oversimplified example to demonstrate these streams:
+
+```bash
+# read the value of foo from stdin
+$ read -r foo
+test
+
+# print the value of var to stdout
+$ echo "value of foo is '$var'"
+value of foo is 'test'
+
+# "echoo" command does not exist so an error message will be printed to stderr
+$ echoo "$var"
+bash: echoo: command not found
+```
+
+These I/O (input/output) streams are typically attached to the user's terminal via [tty](https://linux.die.net/man/4/tty) (TeleTYpe) which can be described as an interface that enables access to the terminal.
+
+As you might have heard multiple times, "everything is a file" according to the UNIX philosophy. This means that the I/O streams should also be a file and this is in fact true:
+
+```bash
+$ file /dev/stdin /dev/stdout /dev/stderr
+
+/dev/stdin:  symbolic link to /proc/self/fd/0
+/dev/stdout: symbolic link to /proc/self/fd/1
+/dev/stderr: symbolic link to /proc/self/fd/2
+```
+
+If you have realized, the file descriptor (unique identifier) of these _abstract_ files are the same as the initial list given above (starts from 0).
+
+<q>
+So if they are files, we should be able to read them right?
+</q>
+
+Actually, no-
+
+<q>
+*types quickly*
+</q>
+
+```bash
+$ cat /dev/stdout
+
+
+
+```
+
+<q>
+Why nothing is happening?
+</q>
+
+\*sigh\* It is because they are not real files, they are just file descriptors linked to TTY or PTY (i.e. emulated TTY AKA PseudoTeletYpe).
+
+```bash
+$ ls -l /proc/self/fd/
+
+lrwx------ - orhun  0 -> /dev/pts/20
+lrwx------ - orhun  1 -> /dev/pts/20
+lrwx------ - orhun  2 -> /dev/pts/20
+```
+
+As you can see, the standard streams are attached to PTYs (i.e. pseudo-terminals) under `/dev/pts`.
+
+<q>Wait, so `/dev/stdout` is a symlink to a file descriptor at `/proc/self/fd/1` and that file descriptor is a symlink to `/dev/pts/20`!? <br><br> What even is `/dev/pts/20` in this case?</q>
+
+```bash
+$ file /dev/pts/20
+
+/dev/pts/20: character special (136/20)
+```
+
+<q>bruh, what?</q>
+
+In Unix, character special files are files that access to I/O devices such as a NULL file (`/dev/null`) and file descriptors. In our case, `stdout` is a file descriptor (1) so it is a character special file. (The name "character special" actually comes from the fact that each character is handled individually.)
+
+Each character special file has a device _major_ number, which identifies the device type, and device _minor_ number, which identifies a specific device of a given device type. So what you see on the right (136/20) means:
+
+```bash
+$ rg '136' /proc/devices
+
+136 pts
+```
+
+- Major 136: a PTS device (`/dev/pts`)
```

Comment:
what is `pts`? just curious

Reply:
emulated TTY AKA PseudoTeletYpe AKA PTYs: <https://en.wikipedia.org/wiki/Pseudoterminal>